ts/nni_manager/core/nniTensorboardManager.ts

```diff
@@ -0,0 +1,208 @@
+import { getLogger, Logger } from '../common/log';
+import { getTunerProc, isAlive, killPid, uniqueString, mkDirPSync, getFreePort } from '../common/utils';
+import { Manager } from '../common/manager';
+import { TensorboardParams, TensorboardTaskStatus, TensorboardTaskInfo, TensorboardManager } from '../common/tensorboardManager';
+
+class TensorboardTaskDetail implements TensorboardTaskInfo {
+    public id: string;
+    public status: TensorboardTaskStatus;
+    public trialJobIdList: string[];
+    public trialLogDirectoryList: string[];
+    public pid?: number;
+    public port?: string;
+
+    constructor(id: string, status: TensorboardTaskStatus, trialJobIdList: string[], trialLogDirectoryList: string[]) {
+        this.id = id;
+        this.status = status;
+        this.trialJobIdList = trialJobIdList;
+        this.trialLogDirectoryList = trialLogDirectoryList;
+    }
+}
+
+class NNITensorboardManager implements TensorboardManager {
+    private log: Logger;
+    private tensorboardTaskMap: Map<string, TensorboardTaskDetail>;
+    private tensorboardVersion: string | undefined;
+    private nniManager: Manager;
+
+    constructor() {
+        this.log = getLogger();
+        this.tensorboardTaskMap = new Map<string, TensorboardTaskDetail>();
+        this.setTensorboardVersion();
+        this.nniManager = component.get(Manager);
+    }
+
+    public async startTensorboardTask(tensorboardParams: TensorboardParams): Promise<TensorboardTaskDetail> {
+        const trialJobIds = tensorboardParams.trials;
+        const trialJobIdList: string[] = [];
+        const trialLogDirectoryList: string[] = [];
+        await Promise.all(trialJobIds.split(',').map(async (trialJobId) => {
+            const trialDataPath = await this.nniManager.getTrialOutputLocalPath(trialJobId);
+            mkDirPSync(trialDataPath);
+            trialJobIdList.push(trialJobId);
+            trialLogDirectoryList.push(trialDataPath);
+        }));
+        this.log.info(`tensorboard: ${trialJobIdList} ${trialLogDirectoryList}`);
+        return this.startTensorboardTaskProcess(trialJobIdList, trialLogDirectoryList);
+    }
+
+    private async startTensorboardTaskProcess(trialJobIdList: string[], trialLogDirectoryList: string[]): Promise<TensorboardTaskDetail> {
+        const host = 'localhost';
+        const port = await getFreePort(host, 6006, 65535);
+        const command = this.getTensorboardStartCommand(trialJobIdList, trialLogDirectoryList, port);
+        this.log.info(`tensorboard start command: ${command}`);
+        const tensorboardProc: ChildProcess = getTunerProc(command, 'ignore', process.cwd(), process.env);
+        const tensorboardTask = new TensorboardTaskDetail(uniqueString(5), 'RUNNING', trialJobIdList, trialLogDirectoryList);
+        this.tensorboardTaskMap.set(tensorboardTask.id, tensorboardTask);
+        tensorboardTask.pid = tensorboardProc.pid;
+        tensorboardTask.port = `${port}`;
```

Comment:
Is it possible that tensorboard is running on another port and this port is incorrect?

Reply:
no, we specified the port, and tensorboard won't automatically find available ports.